src/vulkan/image.cc

```diff
@@ -147,5 +147,82 @@ Result Image::CopyToHost(VkCommandBuffer command) {
   return {};
 }
 
+void Image::ChangeLayout(VkCommandBuffer command,
+                         VkImageLayout old_layout,
+                         VkImageLayout new_layout,
+                         VkPipelineStageFlags from,
+                         VkPipelineStageFlags to) {
+  VkImageMemoryBarrier barrier = {};
+  barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+  barrier.oldLayout = old_layout;
+  barrier.newLayout = new_layout;
+  barrier.srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+  barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+  barrier.image = image_;
+  barrier.subresourceRange = {
+      VK_IMAGE_ASPECT_COLOR_BIT, /* aspectMask */
+      0,                         /* baseMipLevel */
+      1,                         /* levelCount */
+      0,                         /* baseArrayLayer */
+      1,                         /* layerCount */
+  };
+
+  switch (old_layout) {
+    case VK_IMAGE_LAYOUT_PREINITIALIZED:
+      // Based on Vulkan spec, image in VK_IMAGE_LAYOUT_PREINITIALIZED is not
+      // accessible by GPU.
+      barrier.srcAccessMask = VK_ACCESS_HOST_WRITE_BIT;
+      break;
+    case VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL:
+      barrier.srcAccessMask = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+      break;
+    case VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL:
+      barrier.srcAccessMask = VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT;
+      break;
+    case VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL:
+      // An image becomes "transfer dst" only when we send a buffer data to
+      // it.
+      barrier.srcAccessMask = VK_ACCESS_SHADER_READ_BIT |
+                              VK_ACCESS_SHADER_WRITE_BIT |
+                              VK_ACCESS_TRANSFER_WRITE_BIT;
+      break;
+    default:
+      break;
```

Comment:
Let's do barrier.srcAccessMask = 0; just to be clear what's happening.

Reply:
Done.